core/vm/eof_control_flow.go

```diff
@@ -0,0 +1,214 @@
+package vm
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/params"
+)
+
+func validateControlFlow(code []byte, section int, metadata []*functionMetadata, jt *JumpTable) (int, error) {
+	var (
+		maxStackHeight = int(metadata[section].inputs)
+		visitCount     = 0
+		next           = make([]int, 0, 1)
+	)
+	var (
+		stackBoundsMax = make([]uint16, len(code))
+		stackBoundsMin = make([]uint16, len(code))
+	)
+	setBounds := func(pos, min, maxi int) {
+		// The stackboundMax slice is a bit peculiar. We use `0` to denote
+		// not set. Therefore, we use `1` to represent the value `0`, and so on.
+		// So if the caller wants to store `1` as max bound, we internally store it as
+		// `2`.
+		if stackBoundsMax[pos] == 0 { // Not yet set
+			visitCount++
+		}
+		if maxi < 65535 {
+			stackBoundsMax[pos] = uint16(maxi + 1)
+		}
+		stackBoundsMin[pos] = uint16(min)
+		maxStackHeight = max(maxStackHeight, maxi)
+	}
+	getStackMaxMin := func(pos int) (ok bool, min, max int) {
+		maxi := stackBoundsMax[pos]
+		if maxi == 0 { // Not yet set
+			return false, 0, 0
+		}
+		return true, int(stackBoundsMin[pos]), int(maxi - 1)
+	}
+	// set the initial stack bounds
+	setBounds(0, int(metadata[section].inputs), int(metadata[section].inputs))
+
+	qualifiedExit := false
+	for pos := 0; pos < len(code); pos++ {
+		op := OpCode(code[pos])
+		ok, currentStackMin, currentStackMax := getStackMaxMin(pos)
+		if !ok {
+			return 0, errUnreachableCode
+		}
+
+		switch op {
+		case CALLF:
+			arg, _ := parseUint16(code[pos+1:])
+			newSection := metadata[arg]
+			if err := newSection.checkInputs(currentStackMin); err != nil {
+				return 0, fmt.Errorf("%w: at pos %d", err, pos)
+			}
+			if err := newSection.checkStackMax(currentStackMax); err != nil {
+				return 0, fmt.Errorf("%w: at pos %d", err, pos)
+			}
+			delta := newSection.stackDelta()
+			currentStackMax += delta
+			currentStackMin += delta
+		case RETF:
+			if currentStackMax != currentStackMin {
+				return 0, fmt.Errorf("%w: max %d, min %d, at pos %d", errInvalidOutputs, currentStackMax, currentStackMin, pos)
+			}
+			have := int(metadata[section].outputs)
+			if have >= maxOutputItems {
+				return 0, fmt.Errorf("%w: at pos %d", errInvalidNonReturningFlag, pos)
+			}
+			if want := currentStackMin; have != want {
```

Comment:
`have != currentStackMin` (and `have` could be renamed to `outputItems` or something)